Add back Optional for methods

hzsweers opened the pull request:
Nullable is problematic on methods since some return primitive types or void, which is both confusing and something IDEs will lint as errors.

Resolves #306

JakeWharton commented on butterknife-annotations/src/main/java/butterknife/Optional.java:
```diff
@@ -0,0 +1,17 @@
+/**
+ * Denote that the view specified by the injection is not required to be present.
+ * <pre><code>
+ * {@literal @}Optional @OnClick(R.id.subtitle) void onSubtitleClick;
```
You didn't quite complete turning this into a method (semicolon)

hzsweers replied:
Oops, copy pasta mistake since [the original annotation had it](https://github.com/JakeWharton/butterknife/blob/9970b0a88a9f718f7036a71d86686432caea0f6e/butterknife/src/main/java/butterknife/Optional.java#L13)

hzsweers replied:
fed0745

JakeWharton commented on butterknife-compiler/src/main/java/butterknife/compiler/ButterKnifeProcessor.java:
```diff
@@ -1143,17 +1147,19 @@ private String getPackageName(TypeElement type) {
 
   private static boolean isRequiredBinding(Element element) {
-    return !hasAnnotationWithName(element, NULLABLE_ANNOTATION_NAME);
+    return !hasAnnotationWithNames(element, NULLABLE_ANNOTATION_NAME, OPTIONAL_ANNOTATION_NAME);
```
I don't want `@Optional` to work for fields, nor do I want `@Nullable` to work for methods. Each should only work for their respective binding (method vs. field), and only ButterKnife's `@Optional` is supported, not any annotation named "Optional".

hzsweers replied:
So should it just do a manual check here as well to make sure they're valid? Or separate `isOptionalBinding` method (and all that comes with that) entirely? 

I was living it to the annotations themselves, but not actually sure if the compiler will yell at you for using, say, a `@Target(METHOD)` annotation on a field or vice versa.

JakeWharton replied:
I thought separate methods thinking there wouldn't be too much else that had to change. If it truly is a lot, I can take a closer look.

hzsweers replied:
I'll give it a shot and see how it goes

hzsweers replied:
Didn't turn out too bad: 9d77edb

JakeWharton commented:
This mostly looks great, but I want to be a bit more strict in how this annotation is used.

hzsweers commented:
Oof, just saw the checkstyle fail. Weird that it didn't show up for me when I ran connectedCheck locally. Will fix in a bit along with addressing your comments.

JakeWharton commented:
I think it only runs for the 'check' task (which, funny enough, does not
run 'connectedCheck')

On Mon, Oct 12, 2015 at 12:20 AM Zac Sweers <notifications@github.com>
wrote:

> Oof, just saw the checkstyle fail. Weird that it didn't show up for me
> when I ran connectedCheck locally. Will fix in a bit along with addressing
> your comments.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JakeWharton/butterknife/pull/388#issuecomment-147289315>
> .
>

hzsweers commented:
Checkstyle fixed here 2a721d3

JakeWharton commented:
I have squashed some and split some (like the README), fixed the imports, tweaked some language, and changed `@Optional` to only work with the specified annotation instead of any named "Optional".

Thanks for doing this!

hzsweers commented:
:+1: thanks for accepting it! Looking forward to 8.0
